Add memory format support to the `resize_` op.

VitalyFedyunin opened the pull request:
Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29461 [WIP] Experiments
* #29460 [WIP] Switch default of clone, to, *_like ops
* #29459 [WIP] Prepare copy op to switch memory format preservation defaults
* #28991 Adding function to convert Module to channels last
* #28544 Add torch.memory_format support to the TorchScript
* **#28292 Add memory format support to the `resize_` op.**
* #28291 TensroIterator preserve format for binary, ternary operators.
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator

Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

VitalyFedyunin committed:
Add memory format support to the `resize_` op.

Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

[ghstack-poisoned]

VitalyFedyunin commented:
CC @jjsjann123

VitalyFedyunin committed:
Update on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

[ghstack-poisoned]

VitalyFedyunin committed:
Fix rand_like test on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ linter on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

[ghstack-poisoned]

VitalyFedyunin committed:
Fix onnx on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like not simple mapable any more on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like mappable again and supported in jit with memory format on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like now with gradient on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
add TensroIterator tests on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Add some tests on "Add memory format support to the `resize_` op."


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Addressed PR feedback on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Changed model conversion code on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Linter fix on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_ on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix TensorIterator on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix Autograd on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ with new test on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_ on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin commented:
CC @gchanan similar to `resize_as_` PR.

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -5,15 +5,27 @@
   self_->maybe_zero_dim(size.size() == 0);
+  if (optional_memory_format.has_value()) {
+    auto memory_format =
+        optional_memory_format.value_or(MemoryFormat::Contiguous);
```
you already know the tensor has a value, you don't need to use value_or.

gchanan commented on aten/src/ATen/native/ResizeCommon.h:
```diff
@@ -6,16 +6,28 @@
       "). This may be caused by passing a named tensor ",
       "as an `out=` argument; please ensure that the sizes are the same. ");
+  TORCH_CHECK(
+      optional_memory_format.value_or(MemoryFormat::Contiguous) ==
```
shouldn't this just checking if optional_memory_format has a value?

The behavior is we only restride if there's a value, and here we don't restride.

gchanan commented on aten/src/ATen/native/cuda/Resize.cu:
```diff
@@ -4,18 +4,31 @@
   self_->maybe_zero_dim(size.size() == 0);
+  if (optional_memory_format.has_value()) {
+    auto memory_format =
+        optional_memory_format.value_or(MemoryFormat::Contiguous);
```
same as CPU, this should just use value.

gchanan commented on aten/src/ATen/native/quantized/cpu/tensor_operators.cpp:
```diff
@@ -59,7 +59,16 @@ AT_FORALL_OPERATORS(DEFINE_COMPARATOR)
+    Tensor& self,
+    IntArrayRef size,
+    c10::optional<MemoryFormat> optional_memory_format) {
+  auto memory_format =
```
same as the named tensor case, shouldn't this check for not having a value?

Because this behavior differs from non-quantized tensor, where we will actually make something contiguous even if the size matches.

VitalyFedyunin committed:
resize_ PR fixes on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
Add resize_as_ docs on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "Add memory format support to the `resize_` op."

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Allows to simplify patterns like:

1.
	output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});

2.
	output.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	indices.resize_({nbatch, nInputPlane, outputHeight, outputWidth});
	output.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
	indices.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

3.
	gradInput.resize_as_(input);
  	gradInput.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);

Differential Revision: [D18044978](https://our.internmc.facebook.com/intern/diff/D18044978)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/quantized/cpu/tensor_operators.cpp:
```diff
@@ -59,7 +59,14 @@ AT_FORALL_OPERATORS(DEFINE_COMPARATOR)
+    IntArrayRef size,
+    c10::optional<MemoryFormat> optional_memory_format) {
+  TORCH_CHECK(
+      !optional_memory_format.has_value(),
```
are you sure we need this?  quantized tensors should just be able to resized in the same way.

gchanan reviewed (approved):
I'm okay with this but you should get someone from quantization to test/okay the quantization parts.
